### PR TITLE
106 factory implement initialize and create liquidity pool messages

### PR DESCRIPTION
**Pull request by gangov**

### CODE REVIEW

File: contracts/factory/src/storage.rs [hunk 1]
```diff
@@ -1 +1,99 @@
...
+}
+
+#[contracttype]
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct StakeInitInfo {
+    pub stake_wasm_hash: BytesN<32>,
+    pub min_bond: i128,
+    pub max_distributions: u32,
+    pub min_reward: i128,
+}
```
Comment (ueco-jb) [inline contracts/factory/src/storage.rs:47]: Check out: https://github.com/Phoenix-Protocol-Group/phoenix-contracts/issues/113

File: contracts/factory/src/lib.rs [hunk 2]
```diff
@@ -3,6 +3,9 @@ mod contract;
 mod error;
 mod storage;
 
+mod utils;
+
+// lp contract will init token_contract, do we need this one?
```
Comment (ueco-jb) [inline contracts/factory/src/lib.rs:8]: No, we don't.

File: contracts/factory/Makefile [hunk 3]
```diff
@@ -7,6 +7,8 @@ test: build # because of token dependency
 
 build:
 	$(MAKE) -C ../token build || break;
```
Comment (ueco-jb) [inline contracts/factory/Makefile:9]:
```suggestion
```

**Review by ueco-jb — COMMENTED** [review]
Few comments so far.

File: contracts/factory/src/storage.rs (hunk 1 above)
Comment (gangov) [reply contracts/factory/src/storage.rs:47]: started the draft in a separate branch, will keep this here for now and once this is ready will refactor from [#113](https://github.com/Phoenix-Protocol-Group/phoenix-contracts/issues/113)

File: packages/phoenix/src/error.rs
Comment (ueco-jb) [inline packages/phoenix/src/error.rs:1]: Remove that file... We are not getting common error file for all the contracts.

File: packages/phoenix/src/utils.rs [hunk 4]
```diff
@@ -0,0 +1,102 @@
+use decimal::Decimal;
+use soroban_sdk::{contracttype, Address, BytesN};
+
+// Validate if int value is bigger then 0
+#[macro_export]
+macro_rules! validate_int_parameters {
+    ($($arg:expr),*) => {
+        {
+            let mut res: Result<(), $crate::error::ContractError> = Ok(());
```
Comment (ueco-jb) [inline packages/phoenix/src/utils.rs:9]: Panic instead of result and problem solved.

File: contracts/factory/src/contract.rs [hunk 5]
```diff
@@ -11,22 +16,189 @@ pub struct Factory;
...
+            &stake_init_info,
+        );
+
+        let mut lp_vec = get_lp_vec(&env)?;
+
+        lp_vec.push_back(lp_contract_address.clone());
+
+        save_lp_vec(&env, lp_vec);
+
+        env.events().publish(("create", "LP"), &lp_contract_address);
```
Comment (ueco-jb) [inline contracts/factory/src/contract.rs:70]:
```suggestion
        env.events().publish(("create", "liquidity_pool"), &lp_contract_address);
```

File: packages/phoenix/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+//         file = "../../target/wasm32-unknown-unknown/release/phoenix_stake.wasm"
+//     );
+// }
+//
+// #[allow(clippy::too_many_arguments)]
+// pub mod lp_contract {
+//     soroban_sdk::contractimport!(
+//         file = "../../target/wasm32-unknown-unknown/release/phoenix_pair.wasm"
+//     );
+// }
```
Comment (ueco-jb) [inline packages/phoenix/src/lib.rs:24]:
```suggestion
```

File: contracts/factory/src/lib.rs [hunk 7]
```diff
@@ -3,6 +3,15 @@ mod contract;
...
+mod utils;
+
+#[allow(clippy::too_many_arguments)]
+pub mod lp_contract {
+    soroban_sdk::contractimport!(
+        file = "../../target/wasm32-unknown-unknown/release/phoenix_pair.wasm"
+    );
+}
+
 pub mod token_contract {
```
[the remark this answers is not in the document]
Comment (gangov) [reply contracts/factory/src/lib.rs:15]: I'm using that in setup.rs, so it builds everything down the line of the factory contract

**Comment by gangov** [comment]
yes, rebasing fixed %95 of the comments!

File: contracts/factory/src/lib.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (gangov) [reply contracts/factory/src/lib.rs]: fixed

**Comment by gangov** [comment]
We won't be able to add all the `soroban_sdk::contractimport!()` in the Phoenix library because we end up in a circular dependency hell. We can still use `soroban_token_contract.wasm` with that macro, but the build fails when we add `phoenix_stake`. 

Because the Phoenix Library is a dependency of the stake contract during the build of the stake contract it will try to find all the `soroban_sdk::contractimport!()` wasm files. At that time the only present file is the token.wasm so it fails. Tried to disregard the errors during compilation but that requires nightly. Is it okay in the current state?

File: .github/workflows/basic.yml
Comment (ueco-jb) [inline .github/workflows/basic.yml:1]: Remove those changes from this branch and cherry pick it to another. 
Don't do multiple things in one PR.
Create a new one and its purpose will be to compare CI run times.

**Review by ueco-jb — CHANGES_REQUESTED** [review]
Remove changes in CI.
Stop trying to remove token import from all the contracts.

File: contracts/factory/src/lib.rs [hunk 8]
```diff
@@ -3,11 +3,14 @@ mod contract;
 mod error;
 mod storage;
 
-pub mod token_contract {
-    // The import will code generate:
-    // - A ContractClient type that can be used to invoke functions on the contract.
-    // - Any types in the contract that were annotated with #[contracttype].
+mod utils;
+
+#[allow(clippy::too_many_arguments)]
```
Comment (ueco-jb) [inline contracts/factory/src/lib.rs:8]: Clippy was really complaining about this one?

File: packages/phoenix/src/error.rs
Comment (ueco-jb) [inline packages/phoenix/src/error.rs:1]: To remove.

**Review by ueco-jb — APPROVED** [review]
LGTM 
You can remove that error.rs file in a follow up.

**Pull request merged**